Maple2.Server.Game/Model/Field/Actor/ActorState/AiState.cs

```diff
@@ -0,0 +1,453 @@
+
+        skillId = skills[node.Idx - 1].Id;
+        skillLevel = (short)skills[node.Idx - 1].Level;
+
+        SkillRecord? cast = actor.CastSkill(skillId, skillLevel);
+
+        if (cast is null) {
+            actor.AppendDebugMessage($"Failed cast {skillId}[{skillLevel}]\n");
+
+            return;
+        }
+    }
+
+    private void ProcessNode(TeleportNode node) {
+        actor.Position = node.Pos;
+    }
+
+    private void ProcessNode(StandbyNode node) {
+        actor.Navigation?.PathTo(actor.Position); // stop moving
+    }
+
+    private void ProcessNode(SetDataNode node) {
+
+    }
+
+    private void ProcessNode(TargetNode node) {
+
+    }
+
+    private void ProcessNode(SayNode node) {
+        actor.Field.Broadcast(CinematicPacket.BalloonTalk(true, actor.ObjectId, node.Message, node.DurationTick, node.DelayTick));
+
+        actor.AppendDebugMessage($"Say '{node.Message}'");
+    }
+
+    private void ProcessNode(SetValueNode node) {
+
+    }
+
+    private void ProcessNode(ConditionsNode node) {
+        Condition? passed = null;
+
+        foreach (Condition condition in node.Conditions) {
+            if (ProcessCondition((dynamic)condition)) {
+                passed = condition;
+
+                break;
+            }
+        }
+
+        if (passed is null) {
+            actor.AppendDebugMessage("Failed conditions\n");
+
+            return;
+        }
+
+        actor.AppendDebugMessage($"+ Condition: '{passed.Name}'\n");
+
+        Push(passed);
+    }
+
+    private void ProcessNode(JumpNode node) {
+
+    }
+
+    private void ProcessNode(SelectNode node) {
+        var weightedEntries = new WeightedSet<int>();
+
+        bool foundZeroProb = false;
+
+        for (int i = 0; i < node.Prob.Length; ++i) {
+            if (node.Prob[i] == 0) {
+                foundZeroProb = true;
+
+                continue;
+            }
+
+            if (!CanEvaluateNode(node.Entries[i])) {
+                continue;
+            }
+
+            weightedEntries.Add(i, node.Prob[i]);
+        }
+
+        int selectedIndex = -1;
+
+        if (weightedEntries.Count > 0) {
+            int stackIndex = weightedEntries.Get();
+
+            selectedIndex = stackIndex;
+        }
+        else if (foundZeroProb) {
+            // only evaluate if prob=0 nodes are found and none of the weighted ones passed
+            for (int i = 0; i < node.Prob.Length; ++i) {
+                if (i >= node.Entries.Length) {
+                    break;
+                }
+
+                if (node.Prob[i] == 0 && CanEvaluateNode(node.Entries[i])) {
+                    selectedIndex = i;
+
+                    break;
+                }
+            }
+        }
+
+        if (selectedIndex == -1) {
+            return;
+        }
+
+        aiStack[aiStack.Count - 1] = new StackEntry { Node = node, Index = selectedIndex, LockIndex = true };
+
+        Push(node.Entries[selectedIndex]);
+        Process(node.Entries[selectedIndex]);
+    }
```

Comment:
```suggestion
    private void ProcessNode(SelectNode node) {
        var weightedEntries = new WeightedSet<(Entry, int)>();

        for (int i = 0; i < node.Prob.Length; i++) {
            if (!CanEvaluateNode(node.Entries[i])) {
                continue;
            }

            weightedEntries.Add((node.Entries[i], i), node.Prob[i]);
        }

        if (weightedEntries.Count == 0) {
            return;
        }
        
        (Entry entry, int index) selected = weightedEntries.Get();

        aiStack[aiStack.Count - 1] = new StackEntry { Node = node, Index = selected.index, LockIndex = true };

        Push(selected.entry);
        Process(selected.entry);
    }
```

Reply:
The second loop is necessary because the prob=0 nodes aren't weighted, and should be picked as fallbacks. They should not be included in the weighted pick.

Reply:
These 0 nodes will never be picked in a weighted set UNLESS they're the only ones added to the weighted set. This accomplishes that.

Reply:
Ah gotcha